### PR TITLE
WIP: formik and yup

### PR DESCRIPTION
**Pull request by erlenste**
New form with Formik and form validation with Yup.
Input validation on title, event type, starttime, endtime, description, location and image url.
This PR will (maybe) replace the EventForm with NewEventForm.
New components based on the ones used in EventForm has been created to accomodate Formik forms.

### CODE REVIEW

**Comment by erlenste** [comment]
Typescript version was updated from 3.7.2 to 3.8.2 in this PR
Yup requires Typescript version > 3.8 to work

File: src/components/FormikFields/FormikInputField/styled.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import styled from 'styled-components';
+
+export const EventError = styled.div`
```
Comment (sebastto) [inline src/components/FormikFields/FormikInputField/styled.ts:3]: Kan man bruke Formik sin komponent for ErrorMessage?
https://formik.org/docs/api/errormessage

File: src/components/NewEventForm/NewEventForm.tsx [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+function NewEventForm({ onSubmit, ...rest }: EventFormProps) {
+
+  // Declartions
+  const now = new Date();
+  const tomorrow = addDays(startOfDay(now), 1);
+
+  // Utilities
+  const setHours = (date: Date, hours: number) =>
+    set(date, { hours: hours, minutes: 0, seconds: 0, milliseconds: 0 });
+  const parseDate = (date: string, time: string) => new Date(date + 'T' + time);
```
Comment (sebastto) [inline src/components/NewEventForm/NewEventForm.tsx:57]: Er også en handy date string formatter i `src/utils/formatting.ts`

**Review by sebastto — COMMENTED** [review]
Nice validering! 👍
Har også et par småkommentarer, sjekk review

File: src/components/NewEventForm/NewEventForm.tsx [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+
+  // Functions
```
Comment (anhtin) [inline src/components/NewEventForm/NewEventForm.tsx:59]: This comment is incorrect

File: src/components/NewEventForm/NewEventForm.tsx [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+  const eventDate = format(tomorrow, 'yyyy-MM-dd');
+  const eventStartTime = format(setHours(tomorrow, 16), 'HH:mm');
+  const eventEndTime = format(setHours(tomorrow, 17), 'HH:mm');
+
+
+  // Initiation of values
+  const InitialValues = {
```
Comment (anhtin) [inline src/components/NewEventForm/NewEventForm.tsx:66]: This name shouldn't be in PascalCase:
```suggestion
  const initialValues = { 
```

File: src/components/NewEventForm/NewEventForm.tsx [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+    description: rest.description || "",
+    location: rest.location || "", 
+    teamsUrl: rest.teamsUrl || "",
+    imageUrl: rest.imageUrl || "",
+    imageAlt: rest.imageAlt || "",
+    eventStatus: rest.eventStatus || EVENT_STATUS_OPTIONS.Draft
+  };
+  
+  // Validation
+  const EventFormSchema = Yup.object().shape({
```
Comment (anhtin) [inline src/components/NewEventForm/NewEventForm.tsx:81]: This name shouldn't be in PascalCase:
```suggestion
  const eventFormSchema = Yup.object().shape({
```

File: src/components/NewEventForm/NewEventForm.tsx [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+      .string()
+      .max(200, "Location text is too long"),
+    imageUrl: Yup
+      .string()
+      .url("URL not valid"),
+    eventStatus: Yup
+      .string()
+      .oneOf([EVENT_STATUS_OPTIONS.Draft, EVENT_STATUS_OPTIONS.Published])
+      .required('Event status is required')
+  });
```
Comment (anhtin) [inline src/components/NewEventForm/NewEventForm.tsx:124]: This is a constant value, as such, it shouldn't live inside of the component (will cause unnecessary computation on each render)

File: src/components/NewEventForm/NewEventForm.tsx [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+                endTime : parseDate(eventDate, values.endTime),
+                imageUrl : values.imageUrl,
+                imageAlt : values.imageAlt,
+                location : values.location,
+                eventType : values.eventType,
+                eventStatus : values.eventStatus,
+                teamsUrl : values.teamsUrl,    
+              });    
+            } catch (error) {}
+          });
```
Comment (anhtin) [inline src/components/NewEventForm/NewEventForm.tsx:150]: Large anonymous functions like this should be extracted into a named helper function instead. It improves readability 😄

File: src/components/NewEventForm/NewEventForm.tsx [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+          setSubmitting(false);
+        }}
+      >
+        <EventFormWrapper>
+          <h1>{rest.headerTitle}</h1>
+          <Form>
+            <FormikInputField
+              label = "Event name *"
+              name = "name"
+              type = "text"
```
Comment (anhtin) [inline src/components/NewEventForm/NewEventForm.tsx:160]: It's not normal to have a space between attributes/props name and value like this. You should use Prettier if you haven't already installed the plugin.
```suggestion
              label="Event name *"
              name="name"
              type="text"
```
There's more of these below

File: src/components/NewEventForm/NewEventForm.tsx [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+              label = "Image Link"
+              name = "imageUrl"
+              type = "text"
+            />
+            <FormikInputField
+              label = "Image Text"
+              name = "imageAlt"
+              type = "text"
+            />
+            <div style={{justifyContent:'center', display:'flex', alignItems: 'center', marginTop:'1rem'}}>
```
Comment (anhtin) [inline src/components/NewEventForm/NewEventForm.tsx:208]: This styling should be extraced into a styled component. ref. the comment from @sebastto

**Review by anhtin — CHANGES_REQUESTED** [review]
Nice work! There's only some minor stuff that needs to be changed in the name of readbility and easy maintenance 😝 

You might already be thinking of this, but I would just replace all relevant form and fields components with the new version instead of having two separate versions 😛

There's also some repetition in the formik field components that can be extracted into a base component. I'm thinking of `FormikContainer`, `FieldLabel` and `EventError`. The only variable part is really just the input element.

Finally, the `NewEventForm` component is way too large. You should look into how you can split it up into smaller parts to make it easier for a reader to digest the code.

File: src/components/EventForm/EventForm.tsx [hunk 10]
```diff
@@ -99,7 +99,7 @@ function EventForm({ onSubmit, ...rest }: EventFormProps) {
     <Formik
       initialValues={initialValues}
       validationSchema={EventFormSchema}
-      onSubmit={async (values, { setSubmitting }) => {
+      onSubmit={ async (values, { setSubmitting }) => {
```
Comment (anhtin) [inline src/components/EventForm/EventForm.tsx:102]: Unnecessary spacing. You can configure Prettier to be automatically be ran after each save in your editor

File: src/components/EventForm/constants.ts [hunk 11]
```diff
@@ -20,20 +20,41 @@ export const EventFormSchema = Yup.object().shape({
...
-        );
-        return startTimeHour < endTimeHour;
+    .test("Start Time", "Start time cannot be equal, or later than End time", async (value, context) => {
+      const startTimeHour = parseInt(context.parent.startTime.substring(0, 2));
+      const endTimeHour = parseInt(context.parent.endTime.substring(0, 2));
+      const startTimeMinute = parseInt(context.parent.startTime.substring(3, 5));
+      const endTimeMinute = parseInt(context.parent.endTime.toString().substring(3, 5));
+      if (startTimeHour == endTimeHour && endTimeMinute < startTimeMinute) {
+        return false;
       }
```
Comment (anhtin) [inline src/components/EventForm/constants.ts]: You should use `date-fns` here instead

File: src/components/EventForm/constants.ts [hunk 12]
```diff
@@ -20,20 +20,41 @@ export const EventFormSchema = Yup.object().shape({
...
+      const startTimeHour = parseInt(context.parent.startTime.substring(0, 2));
+      const endTimeHour = parseInt(context.parent.endTime.substring(0, 2));
+      const startTimeMinute = parseInt(context.parent.startTime.substring(3, 5));
+      const endTimeMinute = parseInt(context.parent.endTime.toString().substring(3, 5));
+      if (startTimeHour == endTimeHour && endTimeMinute < startTimeMinute) {
+        return false;
+      }
+      if (startTimeHour > endTimeHour) {
+        return false;
+      }
```
Comment (anhtin) [inline src/components/EventForm/constants.ts:52]: Same as other comment: you should use `date-fns` here

**Comment by erlenste** [comment]
Changes from 2nd review are implemented

**Pull request merged**